File: app/src/main/java/com/estasvegano/android/estasvegano/model/ProductModel.java
package com.estasvegano.android.estasvegano.model;

import android.graphics.Bitmap;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import com.estasvegano.android.estasvegano.data.web.EVeganoApi;
import com.estasvegano.android.estasvegano.data.web.ErrorParser;
import com.estasvegano.android.estasvegano.data.web.request.AddProductRequest;
import com.estasvegano.android.estasvegano.entity.Complain;
import com.estasvegano.android.estasvegano.entity.Photo;
import com.estasvegano.android.estasvegano.entity.Product;
import com.estasvegano.android.estasvegano.entity.ProductType;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.ByteArrayOutputStream;

import okhttp3.MediaType;
import okhttp3.RequestBody;

// TODO: add caching
public class ProductModel {

    @NonNull
    private final EVeganoApi api;

    @NonNull
    private final ErrorParser errorParser;

    @NonNull
    private final ObjectMapper objectMapper;

    public ProductModel(@NonNull EVeganoApi api, @NonNull ErrorParser errorParser, @NonNull ObjectMapper objectMapper) {
        this.api = api;
        this.errorParser = errorParser;
        this.objectMapper = objectMapper;
    }

    @CheckResult
    @NonNull
    public rx.Single<Product> checkCode(String code, String codeType) {
        return api.checkCode(code, codeType)
                .flatMap(errorParser::checkIfError)
                .map(data -> objectMapper.convertValue(data, Product.class));
    }

    @CheckResult
    @NonNull
    public rx.Single<Product> addProduct(
            @NonNull String title,
            @NonNull ProductType type,
            @NonNull String code,
            @NonNull String format,
            long categoryId,
            long producerId
    ) {
        return api.addProduct(
                AddProductRequest.builder()
                        .title(title)
                        .info(type)
                        .categoryId(categoryId)
                        .producerId(producerId)
                        .code(code)
                        .codeType(format)
                        .build()
        )
                .flatMap(errorParser::checkIfError)
                .map(data -> objectMapper.convertValue(data, Product.class));
    }

    @CheckResult
    @NonNull
    public rx.Single<Photo> uploadPhoto(long productId, Bitmap image) {
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        image.compress(Bitmap.CompressFormat.PNG, 100, stream);
        byte[] imageBytes = stream.toByteArray();
        RequestBody requestBody = RequestBody.create(MediaType.parse("image/jpeg"), imageBytes);
        return api.uploadPhoto(productId, requestBody)
                .flatMap(errorParser::checkIfError)
                .map(data -> objectMapper.convertValue(data, Photo.class));
    }

    @CheckResult
    @NonNull
    public rx.Single<Void> complain(long productId, Complain complain) {
        return api.complain(productId, complain).flatMap(errorParser::checkIfError);
    }
}

File: app/src/main/java/com/estasvegano/android/estasvegano/model/ProductModel.kt
package com.estasvegano.android.estasvegano.model

import android.graphics.Bitmap
import android.support.annotation.CheckResult
import com.estasvegano.android.estasvegano.data.web.EVeganoApi
import com.estasvegano.android.estasvegano.data.web.ErrorParser
import com.estasvegano.android.estasvegano.data.web.request.AddProductRequest
import com.estasvegano.android.estasvegano.entity.Complain
import com.estasvegano.android.estasvegano.entity.Photo
import com.estasvegano.android.estasvegano.entity.Product
import com.estasvegano.android.estasvegano.entity.ProductType
import com.fasterxml.jackson.databind.ObjectMapper
import io.reactivex.Single
import okhttp3.MediaType
import okhttp3.RequestBody
import java.io.ByteArrayOutputStream

// TODO: add caching
class ProductModel(private val api: EVeganoApi, private val errorParser: ErrorParser, private val objectMapper: ObjectMapper) {

    @CheckResult
    fun checkCode(code: String, codeType: String): Single<Product> {
        return api.checkCode(code, codeType)
                .flatMap { errorParser.checkIfError(it) }
                .map { objectMapper.convertValue(it, Product::class.java) }
    }

    @CheckResult
    fun addProduct(
            title: String,
            type: ProductType,
            code: String,
            format: String,
            categoryId: Long,
            producerId: Long
    ): Single<Product> {
        return api.addProduct(
                AddProductRequest(
                        title = title,
                        info = type,
                        categoryId = categoryId,
                        producerId = producerId,
                        code = code,
                        codeType = format
                )
        )
                .flatMap { errorParser.checkIfError(it) }
                .map { objectMapper.convertValue(it, Product::class.java) }
    }

    @CheckResult
    fun uploadPhoto(productId: Long, image: Bitmap): Single<Photo> {
        val stream = ByteArrayOutputStream()
        image.compress(Bitmap.CompressFormat.PNG, 100, stream)
        val imageBytes = stream.toByteArray()
        val requestBody = RequestBody.create(MediaType.parse("image/jpeg"), imageBytes)
        return api.uploadPhoto(productId, requestBody)
                .flatMap { errorParser.checkIfError(it) }
                .map { objectMapper.convertValue(it, Photo::class.java) }
    }

    @CheckResult
    fun complain(productId: Long, complain: Complain): Single<Any> {
        return api.complain(productId, complain)
                .flatMap { errorParser.checkIfError(it) }
    }
}
